packages/flutter/lib/src/painting/path_border.dart

```diff
@@ -0,0 +1,191 @@
+    @required this.pathBuilder,
+    BorderSide border = const BorderSide(),
+  })  : assert(pathBuilder != null),
+        assert(border != null),
+        _border = border,
+        pathPaint = border.toPaint();
+
+  /// Cretes a [PathBorder] with the specified paint.
+  ///
+  /// This constructor is useful if you wish to directly control the [Paint]
+  /// used to paint the border, such as applying a gradient or setting
+  /// [Paint.strokeCap], [Paint.strokeJoin], or [Paint.strokeMiterLimit] to a
+  /// non-default value.
+  ///
+  /// The [pathPaint] must have a [PaintingStyle.stroke] style. The fill color
+  /// for a [BoxDecoration] is specified by [BoxDecoration.color].
+  ///
+  /// The [pathPaint] and [pathBuilder] properties must not be null.
+  PathBorder.withPaint({
+    @required this.pathBuilder,
+    @required this.pathPaint,
+  })  : assert(pathBuilder != null),
+        assert(pathPaint != null),
+        assert(pathPaint.style == PaintingStyle.stroke, 'PathBorder does not support drawing a border with a fill.'),
+        _border = BorderSide(
+          color: pathPaint.color,
+          width: pathPaint.strokeWidth,
+          style: BorderStyle.solid,
+        );
+
+  /// The callback to use when creating a path in a given [Rect].
+  ///
+  /// This callback fire at paint time with the boundary and [TextDirection] for
+  /// the border. The implementation is free to return any path it wishes. No
+  /// transformations will be applied to the resulting path, which will be
+  /// rendered on the same canvas as the box it is decorating. If it draws
+  /// outside of the boundary, it may or may not be visible as larger than the
+  /// box. Alternatively , if it draws a path that is much smaller than the
+  /// boundary, it will draw inside the decorated box rather than at the edges.
+  ///
+  /// {@tool sample}
+  /// The [Path] class has several methods which take a [Rect] as a parameter,
+  /// which are useful in implementing this callback.
+  ///
+  /// ```dart
+  /// (Rect boundary, TextDirection textDirection) {
+  ///   return Path()..addRRect(
+  ///     RRect.fromRectAndRadius(
+  ///       boundary,
+  ///       Radius.circular(5.0),
+  ///     ),
+  ///   );
+  /// }
+  /// ```
+  /// {@end-tool}
+  final PathBorderBuilder pathBuilder;
+
+  /// The [Paint] used to draw the path border.
+  ///
+  /// This paint must have its [Paint.style] set to [PaintingStyle.stroke]. To
+  /// set the fill on a [BoxDecoration], use the [BoxDecoration.color] property.
+  final Paint pathPaint;
+
+  @override
+  bool get isUniform => true;
+
+  final BorderSide _border;
+
+  @override
+  BorderSide get bottom => _border;
+
+  @override
+  BorderSide get top => _border;
+
+  @override
+  Path getInnerPath(Rect rect, {TextDirection textDirection}) {
+    return pathBuilder(rect, textDirection);
+  }
+
+  @override
+  Path getOuterPath(Rect rect, {TextDirection textDirection}) {
+    return pathBuilder(rect, textDirection);
+  }
```

Comment:
inner and outer being the same seems dubious

Reply:
We have no idea what the path is - just that we expect it to be stroked. It may not be closed.

Reply:
That seems to me to violate the Shape contract.